dll/win32/shell32/dialogs/folder_options.cpp

```diff
@@ -1525,6 +1610,130 @@ FindSelectedItem(
     return NULL;
 }
 
+INT_PTR CALLBACK
+NewExtDlgProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    /* FIXME */
+    static HWND s_hListView = NULL;
+    switch (uMsg)
+    {
+        case WM_INITDIALOG:
+            s_hListView = (HWND)lParam;
+            return TRUE;
+
+        case WM_COMMAND:
+            switch (LOWORD(wParam))
+            {
+                case IDOK:
+                    EndDialog(hwndDlg, IDOK);
+                    break;
+                case IDCANCEL:
+                    EndDialog(hwndDlg, IDCANCEL);
+                    break;
+            }
+            break;
+    }
+    UNREFERENCED_LOCAL_VARIABLE(s_hListView);
+    return 0;
+}
+
+void FileTypesDlg_OnNew(HWND hwndDlg, HWND hListView)
+{
+    DialogBoxParamW(shell32_hInstance, MAKEINTRESOURCEW(IDD_NEW_EXT),
+                    hwndDlg, NewExtDlgProc, (LPARAM)hListView);
+}
+
+void FileTypesDlg_OnChange(HWND hwndDlg, HWND hListView)
+{
+    OPENASINFO Info;
+    PFOLDER_FILE_TYPE_ENTRY pItem = FindSelectedItem(hListView);
+    if (pItem)
+    {
+        Info.oaifInFlags = OAIF_ALLOW_REGISTRATION | OAIF_REGISTER_EXT;
+        Info.pcszClass = pItem->FileExtension;
+        SHOpenWithDialog(hwndDlg, &Info);
+    }
+}
+
+void FileTypesDlg_OnDelete(HWND hwndDlg, HWND hListView)
+{
+    const UINT Flags = LVNI_ALL | LVNI_SELECTED;
+    INT iItem = ListView_GetNextItem(hListView, -1, Flags);
+    if (iItem == -1)
+        return;
+
+    PFOLDER_FILE_TYPE_ENTRY pItem = FindSelectedItem(hListView, iItem);
+    if (pItem == NULL)
+        return;
+
+    // FIXME: registry
```

Comment:
What needs to be done in addition?

Reply:
HKEY_CLASSES_ROOT contains many file type informations. For example, there is ".txt" registry key. And "txtfile" registry key also exists. If the user had requested deletion, we have to delete both of ".txt" and "txtfile".

Reply:
Implemented file type deletion.